### PR TITLE
Data base

### PR DESCRIPTION
**Pull request by Kaisar283**
@Alexmansar Hi
I've done the task, please check

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -24,12 +38,14 @@ jobs:
           distribution: 'temurin'
           cache: maven
 
-      - name: Run the Maven verify phase
-        run: mvn --batch-mode --update-snapshots verify
+      - name: Build the project (without tests)
```
Comment (Alexmansar) [inline .github/workflows/ci.yaml:41]: why we skip tests?

File: src/main/java/kz/andersen/java_intensive_13/models/User.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import kz.andersen.java_intensive_13.enums.UserRole;
+
+import java.time.ZonedDateTime;
+
+public class User {
+
+    private Long id;
+    private String fistName;
+    private String lastName;
+    UserRole userRole;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/models/User.java]: what about access modifier?

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 3]
```diff
@@ -20,21 +32,85 @@ public static ApartmentStorage getInstance(){
...
+        double price = apartment.getPrice();
+        boolean isReserved = apartment.getIsReserved();
+        ZonedDateTime createdAt = ZonedDateTime.now(ZoneId.systemDefault());
+        ZonedDateTime updatedAt = ZonedDateTime.now(ZoneId.systemDefault());
+
+        String SQL_QUERY = """
+                INSERT INTO public.apartment(
+                	id, price, "isReserved", "reservedBy", created_at, updated_at)
+                	VALUES (?, ?, ?, ?, ?, ?);
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: use queries as a constants on the class level

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 4]
```diff
@@ -20,21 +32,85 @@ public static ApartmentStorage getInstance(){
...
+        boolean isReserved = apartment.getIsReserved();
+        long userId = apartment.getReservedBy().getId();
+        ZonedDateTime createdAt = apartment.getCreatedAt();
+        ZonedDateTime updatedAt = apartment.getUpdatedAt();
+
+        String SQL_QUERY = String.format("""
+                UPDATE public.apartment
+                	SET id=?, price=?, "isReserved"=?, "reservedBy"=?, created_at=?, updated_at=?
+                	WHERE id=%d;
+                """, apartment.getId() );
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: all in general

File: src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    public int saveUser(User user){
+        long userId = user.getId();
+        String firstName = user.getFistName();
+        String lastName = user.getLastName();
+        String userRole = user.getUserRole().toString();
+        ZonedDateTime createdAt = ZonedDateTime.now(ZoneId.systemDefault());
+        ZonedDateTime updatedAt = ZonedDateTime.now(ZoneId.systemDefault());
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java]: you ca use it directly in lines 53-58

File: src/main/java/kz/andersen/java_intensive_13/db_config/DataSource.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import java.sql.SQLException;
+
+public class DataSource {
+    private static HikariConfig config = new HikariConfig();
+    private static HikariDataSource ds;
+
+    static {
+        config.setJdbcUrl( "jdbc:postgresql://localhost:5432/hotel-db" );
+        config.setUsername( "postgres" );
+        config.setPassword( "admin" );
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/db_config/DataSource.java]: can we read it from file?

**Review by Alexmansar — CHANGES_REQUESTED** [review]
@Kaisar283 Hi) I have some general comments, so please fix it in each case)

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (Kaisar283) [reply .github/workflows/ci.yaml:41]: All tests are run after migration to the database. Liquibase does not migrate to the database when building the project. So I separate test phase. Tests are run with DB connection

Comment (Alexmansar) [reply .github/workflows/ci.yaml:41]: ok)

**Comment by Kaisar283** [comment]
@Alexmansar please check

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 7]
```diff
@@ -20,21 +44,76 @@ public static ApartmentStorage getInstance(){
...
+        }
     }
 
     /**
      * @return the original Apartments list
      */
     public List<Apartment> getApartments(){
-        return apartments;
+        List<Apartment> apartmentList = new ArrayList<>();
+        String SQLQuery = "SELECT * FROM apartment LEFT JOIN public.user ON apartment.\"reservedBy\" = public.user.id;";
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: and this

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 8]
```diff
@@ -45,48 +124,140 @@ public List<Apartment> getApartments(){
...
      */
     public Optional<Apartment> getApartmentById(int apartmentId){
-        for (Apartment apartment:apartments) {
-            if(apartment.getId() == apartmentId){
-                return Optional.of(apartment);
+        String SQLQuery = String.format("""
+                SELECT * FROM apartment LEFT JOIN public.user
+                ON apartment.\"reservedBy\" = public.user.id
+                WHERE apartment.id = '%d';
+                """, apartmentId);
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 9]
```diff
@@ -45,48 +124,140 @@ public List<Apartment> getApartments(){
...
      */
     public List<Apartment> sortApartmentByPrice(){
-        List<Apartment> apartmentList = new ArrayList<>(Collections.unmodifiableList(apartments));
-        apartmentList.sort(Comparator.comparingDouble(Apartment::getPrice));
+        List<Apartment> apartmentList = new ArrayList<>();
+        String SQLQuery = """
+                SELECT * FROM apartment LEFT JOIN public.user
+                	ON apartment."reservedBy" = public.user.id
+                	ORDER BY apartment.price DESC;
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 10]
```diff
@@ -45,48 +124,140 @@ public List<Apartment> getApartments(){
...
      */
     public List<Apartment> sortApartmentById(){
-        List<Apartment> apartmentList = new ArrayList<>(Collections.unmodifiableList(apartments));
-        apartmentList.sort(Comparator.comparingInt(Apartment::getId));
+        List<Apartment> apartmentList = new ArrayList<>();
+        String SQLQuery = """
+                SELECT * FROM apartment LEFT JOIN public.user
+                	ON apartment."reservedBy" = public.user.id
+                	ORDER BY apartment.id ASC;
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 11]
```diff
@@ -45,48 +124,140 @@ public List<Apartment> getApartments(){
...
-        apartments.sort(Comparator.comparing(
-                apartment -> apartment.getReservedBy() == null ? "" : apartment.getReservedBy().getName()
-        ));
-        return apartments.reversed();
+        List<Apartment> apartmentList = new ArrayList<>();
+        String SQLQuery = """
+                SELECT * FROM apartment LEFT JOIN public.user
+                	ON apartment."reservedBy" = public.user.id
+                	ORDER BY public.user.first_name ASC;
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 12]
```diff
@@ -45,48 +124,140 @@ public List<Apartment> getApartments(){
...
      */
     public List<Apartment> sortedApartmentByReservationStatus(){
-        apartments.sort(Comparator.comparing(Apartment::getIsReserved));
-        return apartments.reversed();
+        List<Apartment> apartmentList = new ArrayList<>();
+        String SQLQuery = """
+                SELECT * FROM apartment LEFT JOIN public.user
+                	ON apartment."reservedBy" = public.user.id
+                    ORDER BY apartment."isReserved" DESC;
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Optional;
+
+import static java.sql.Types.BIGINT;
+
+public class UserRepository {
+
+    public Optional<User> findUserById(long userId){
+        String SQLQuery = String.format("""
+                SELECT * FROM public."user" WHERE "user".id = '%d';
+                """, userId);
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java]: this

File: src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    public int saveUser(User user){
+        String SQL_QUERY = """
+                INSERT INTO public."user"(
+                	id, first_name, last_name, user_role, created_at, updated_at)
+                	VALUES (?, ?, ?, ?, ?, ?);
+                """;
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java]: this

**Review by Alexmansar — CHANGES_REQUESTED** [review]

**Comment by Kaisar283** [comment]
@Alexmansar please check

File: src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java [hunk 15]
```diff
@@ -12,18 +11,24 @@
...
-        String SQLQuery = String.format("""
-                SELECT * FROM public."user" WHERE "user".id = '%d';
-                """, userId);
-        try(Connection connection = DataSource.getConnection();
-            PreparedStatement pst = connection.prepareStatement(SQLQuery);
-            ResultSet results = pst.executeQuery();
-        ){
+    private final String SELECT_BY_ID_QUERY = "SELECT * FROM public.\"user\" WHERE \"user\".id = '%d'";
+
+    private final String INSET_USER_QUERY = """
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/UserRepository.java]: INSERT

File: src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java [hunk 16]
```diff
@@ -19,25 +19,68 @@ public class ApartmentStorage {
     private static volatile ApartmentStorage instance;
     private DataSource dataSource;
 
-    private final String INSET_INTO_APARTMENT_QUERY = """
-                INSERT INTO public.apartment(
-                	id, price, "isReserved", "reservedBy", created_at, updated_at)
-                	VALUES (?, ?, ?, ?, ?, ?);
-                """;
+    private final String INSET_APARTMENT_QUERY = """
```
Comment (Alexmansar) [inline src/main/java/kz/andersen/java_intensive_13/repository/ApartmentStorage.java]: same

**Review by Alexmansar — APPROVED** [review]
@Kaisar283 Please, fix small remarks and merge

**Pull request merged**